# Spam SMS Detection

This project classifies SMS messages as Spam or Ham using Machine Learning.

Internship: CodSoft Machine Learning Internship

Import libraries


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

print("Libraries imported successfully!")

Libraries imported successfully!


Upload Kaggle dataset

In [3]:
from google.colab import files

uploaded = files.upload()

Saving archive (5).zip to archive (5) (1).zip


Unzip the file

In [8]:
import zipfile

with zipfile.ZipFile("archive (5) (1).zip", "r") as zip_ref:
    zip_ref.extractall()

print("Dataset extracted successfully!")

Dataset extracted successfully!


Check Files

In [9]:
import os
print(os.listdir())

['.config', 'archive (5) (1).zip', 'spam.csv', 'archive (5).zip', 'sample_data']


Load

In [10]:
df = pd.read_csv("spam.csv", encoding="latin-1")

Keep only useful columns

In [11]:
df = df[['v1', 'v2']]

df.columns = ['label', 'message']

df.head()

,label,message
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


Check Dataset

In [12]:
print("Rows and Columns:")
print(df.shape)

print("\nMissing Values:")
print(df.isnull().sum())

print("\nSpam vs Ham:")
print(df['label'].value_counts())

Rows and Columns:
(5572, 2)

Missing Values:
label      0
message    0
dtype: int64

Spam vs Ham:
label
ham     4825
spam     747
Name: count, dtype: int64


Convert Labels into Numbers

In [13]:
df['label'] = df['label'].map({
    'ham': 0,
    'spam': 1
})

df.head()

,label,message
0,0,"Go until jurong point, crazy.. Available only ..."
1,0,Ok lar... Joking wif u oni...
2,1,Free entry in 2 a wkly comp to win FA Cup fina...
3,0,U dun say so early hor... U c already then say...
4,0,"Nah I don't think he goes to usf, he lives aro..."


Prepare Features and Labels

In [14]:
X = df['message']
y = df['label']

print(X.head())

print("\n")

print(y.head())

0    Go until jurong point, crazy.. Available only ...
1                        Ok lar... Joking wif u oni...
2    Free entry in 2 a wkly comp to win FA Cup fina...
3    U dun say so early hor... U c already then say...
4    Nah I don't think he goes to usf, he lives aro...
Name: message, dtype: object


0    0
1    0
2    1
3    0
4    0
Name: label, dtype: int64


Convert Text into Numbers (TF-IDF)

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(
    stop_words='english',
    max_features=5000
)

X_tfidf = tfidf.fit_transform(X)

print(X_tfidf.shape)

(5572, 5000)


Train-Test Split

In [16]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_tfidf,
    y,
    test_size=0.2,
    random_state=42
)

print(X_train.shape)
print(X_test.shape)

(4457, 5000)
(1115, 5000)


Logistic Regression

In [17]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

lr_model = LogisticRegression(max_iter=1000)

lr_model.fit(X_train, y_train)

lr_pred = lr_model.predict(X_test)

lr_accuracy = accuracy_score(y_test, lr_pred)

print("Logistic Regression Accuracy:", lr_accuracy)

Logistic Regression Accuracy: 0.9488789237668162


Naive Bayes

In [18]:
from sklearn.naive_bayes import MultinomialNB

nb_model = MultinomialNB()

nb_model.fit(X_train, y_train)

nb_pred = nb_model.predict(X_test)

nb_accuracy = accuracy_score(y_test, nb_pred)

print("Naive Bayes Accuracy:", nb_accuracy)

Naive Bayes Accuracy: 0.9748878923766816


SVM

In [19]:
from sklearn.svm import LinearSVC

svm_model = LinearSVC()

svm_model.fit(X_train, y_train)

svm_pred = svm_model.predict(X_test)

svm_accuracy = accuracy_score(y_test, svm_pred)

print("SVM Accuracy:", svm_accuracy)

SVM Accuracy: 0.9739910313901345


Compare Models

In [20]:
print("Logistic Regression Accuracy:", lr_accuracy)
print("Naive Bayes Accuracy:", nb_accuracy)
print("SVM Accuracy:", svm_accuracy)

Logistic Regression Accuracy: 0.9488789237668162
Naive Bayes Accuracy: 0.9748878923766816
SVM Accuracy: 0.9739910313901345


Best Model

In [21]:
model_accuracies = {
    "Logistic Regression": lr_accuracy,
    "Naive Bayes": nb_accuracy,
    "SVM": svm_accuracy
}

best_model_name = max(model_accuracies, key=model_accuracies.get)
best_accuracy = model_accuracies[best_model_name]

print("Best Model:", best_model_name)
print("Best Accuracy:", best_accuracy)

Best Model: Naive Bayes
Best Accuracy: 0.9748878923766816


Save Model

In [22]:
import joblib

joblib.dump(nb_model, "spam_sms_model.pkl")
joblib.dump(tfidf, "tfidf_vectorizer.pkl")

print("Model saved successfully!")

Model saved successfully!


Download

In [23]:
from google.colab import files

files.download("spam_sms_model.pkl")
files.download("tfidf_vectorizer.pkl")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Test_1

In [24]:
sample_sms = ["Congratulations! You have won a free lottery prize. Click here to claim now."]

sample_tfidf = tfidf.transform(sample_sms)

prediction = nb_model.predict(sample_tfidf)

if prediction[0] == 1:
    print("Spam Message")
else:
    print("Ham Message")

Spam Message


Test_2

In [25]:
sample_sms = ["Hey, are you coming to class tomorrow?"]

sample_tfidf = tfidf.transform(sample_sms)

prediction = nb_model.predict(sample_tfidf)

if prediction[0] == 1:
    print("Spam Message")
else:
    print("Ham Message")

Ham Message


Test_3

In [29]:
sample_sms = ["Hey, I have reached home safely. Thanks for helping me today!"]
sample_tfidf = tfidf.transform(sample_sms)

prediction = nb_model.predict(sample_tfidf)

if prediction[0] == 1:
    print("Spam Message")
else:
    print("Ham Message")

Ham Message


**CONCLUSION**:

This project classifies SMS messages as Spam or Ham using Machine Learning and TF-IDF vectorization.

Three machine learning models were trained and compared:
1. Logistic Regression
2. Naive Bayes
3. Support Vector Machine (SVM)

The Naive Bayes model achieved the highest accuracy of 97.49%, outperforming the other models. The final model can successfully identify spam messages and help filter unwanted SMS communications.